[MINOR][SQL] Refactor RowEncoder to use existing (De)serializerBuildHelper methods

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

This patch proposes to reuse existing methods in (De)serializerBuildHelper in RowEncoder to achieve deduplication as well as consistent creation of serialization/deserialization of same type.

## How was this patch tested?

Existing UT.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #103169 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103169/testReport)** for PR 24014 at commit [`c72b3c9`](https://github.com/apache/spark/commit/c72b3c9bbd7f6386312db0a8e05db0d2fe36c051).

SparkQA commented:
**[Test build #103169 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103169/testReport)** for PR 24014 at commit [`c72b3c9`](https://github.com/apache/spark/commit/c72b3c9bbd7f6386312db0a8e05db0d2fe36c051).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103169/
Test PASSed.

HeartSaVioR commented:
cc. @cloud-fan @maropu since we've reviewed similar improvement PRs before.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -22,7 +22,9 @@ import scala.reflect.ClassTag
+import org.apache.spark.sql.catalyst.SerializerBuildHelper._
 import org.apache.spark.sql.catalyst.analysis.GetColumnByOrdinal
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.objects._
```
I saw some unused imports here, so (not related to this pr though...) could you remove them?

HeartSaVioR replied:
Sure! I'll clean out them.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SerializerBuildHelper.scala:
```diff
@@ -188,11 +240,45 @@ object SerializerBuildHelper {
-    val obj = NewInstance(udtClass, Nil, dataType = ObjectType(udtClass))
-    Invoke(obj, "serialize", udt, inputObject :: Nil)
+      udtClass: Class[_],
+      propagateNull: Boolean = true,
```
In `SerializerBuildHelper`/`DeserializerBuildHelper`, some functions have no argument for `nullable` and, on the other hand, the other functions have an argument for `nullable`. Is there any consistent policy for the argument here?

HeartSaVioR replied:
That's because of [SPARK-26991](https://issues.apache.org/jira/browse/SPARK-26991) - the method doesn't have argument if the value of `nullable` from both of Scala and Java side are same, have argument otherwise.

If we want to be consistent, we can add `nullable` to all methods consistently: ideal situation would be like we investigate SPARK-26991 successfully, and remove that argument and use one way consistently between Scala and Java side. I just failed to investigate SPARK-26991 so can't do that.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -87,142 +89,50 @@ object RowEncoder {
-        case _ => MapObjects(
+          createSerializerForPrimitiveArrayViaArrayData(inputObject, t)
+
+        case _ => createSerializerForMapObjects(inputObject, ObjectType(classOf[Object]),
```
nit: style
```
        case _ =>
          createSerializerForMapObjects(
            inputObject,
            ObjectType(classOf[Object]),
            element => {
              val value = serializerFor(ValidateExternalType(element, et), et)
              expressionWithNullSafety(value, containsNull, WalkedTypePath())
            })
```

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -285,63 +195,27 @@ object RowEncoder {
-        "microsToInstant",
-        input :: Nil,
-        returnNullable = false)
+    case TimestampType if SQLConf.get.datetimeJava8ApiEnabled => createDeserializerForInstant(input)
```
How about this?
```

    case TimestampType =>
      if (SQLConf.get.datetimeJava8ApiEnabled) {
        createDeserializerForInstant(input)
      } else {
        createDeserializerForSqlTimestamp(input)
      }
```

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -285,63 +195,27 @@ object RowEncoder {
-        "daysToLocalDate",
-        input :: Nil,
-        returnNullable = false)
+    case DateType if SQLConf.get.datetimeJava8ApiEnabled => createDeserializerForLocalDate(input)
```
ditto

SparkQA commented:
**[Test build #103313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103313/testReport)** for PR 24014 at commit [`83e06a6`](https://github.com/apache/spark/commit/83e06a612ac5a69e6a9db9ee6ef548805b247169).

SparkQA commented:
**[Test build #103313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103313/testReport)** for PR 24014 at commit [`83e06a6`](https://github.com/apache/spark/commit/83e06a612ac5a69e6a9db9ee6ef548805b247169).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103313/
Test PASSed.

cloud-fan commented:
The refactor was to make `JavaTypeInference` and `ScalaReflection` consistent. what's the benefits to do it for `RowEncoder`?

HeartSaVioR commented:
While I'm adding new methods to `SerializerBuildHelper` / `DeserializerBuildHelper`, I've also found some parts which are exactly doing same things for what `JavaTypeInference` and `ScalaReflection` are doing. e.g. `createDeserializerForJavaBigDecimal`, `createDeserializerForString`, `createSerializerForDecimalType`, etc. 

So for some parts they're doing different, but some other parts they're just doing same. I can remove added methods and just let RowEncoder reuse available methods in `SerializerBuildHelper` / `DeserializerBuildHelper` if we are not seeing benefit of doing it. Please let me know your thought about it.

cloud-fan commented:
> I can remove added methods and just let RowEncoder reuse available methods

This SGTM. Look at the code changes seems only a few methods can be reused.

HeartSaVioR committed:
[MINOR][SQL] Refactor RowEncoder to use existing (De)serializerBuildHelper methods

HeartSaVioR committed:
Further extract method for deduplications

HeartSaVioR committed:
fix nits

SparkQA commented:
**[Test build #103360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103360/testReport)** for PR 24014 at commit [`d2842dd`](https://github.com/apache/spark/commit/d2842ddbf31e4eacd896a2f158a5228ce06ea2b8).

HeartSaVioR commented:
Just changed to reuse existing methods in (De)serializerBuildHelper in RowEncoder. Minor extracting method is also done.

Also updated title and content of PR. I'll resolve SPARK-27092 as invalid.

SparkQA commented:
**[Test build #103361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103361/testReport)** for PR 24014 at commit [`2c665fb`](https://github.com/apache/spark/commit/2c665fbdacda14881593bcfdb4d194d5a6858dcf).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103360/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103361/
Test FAILed.

HeartSaVioR commented:
Retest this, please

SparkQA commented:
**[Test build #103362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103362/testReport)** for PR 24014 at commit [`2c665fb`](https://github.com/apache/spark/commit/2c665fbdacda14881593bcfdb4d194d5a6858dcf).

SparkQA commented:
**[Test build #103363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103363/testReport)** for PR 24014 at commit [`2c665fb`](https://github.com/apache/spark/commit/2c665fbdacda14881593bcfdb4d194d5a6858dcf).

SparkQA commented:
**[Test build #103362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103362/testReport)** for PR 24014 at commit [`2c665fb`](https://github.com/apache/spark/commit/2c665fbdacda14881593bcfdb4d194d5a6858dcf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #103363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103363/testReport)** for PR 24014 at commit [`2c665fb`](https://github.com/apache/spark/commit/2c665fbdacda14881593bcfdb4d194d5a6858dcf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103362/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103363/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

HeartSaVioR commented:
Thanks all for reviewing and merging!
